### ISSUE TITLE
buildcache create: S3 push fails when binary already exists at mirror w/ same dag hash, different full hash

### ISSUE THREAD

**Issue by eugeneswalker** [spack/spack#14346]
Using branch `develop` at commit `7a97dc3770d` , I have a `spec.yaml` for a package that needs updating at the build cache due to differing full hashes, where the DAG hash is the same:

```
[lpeyrala@login1.ascent ~]$ spack buildcache check --rebuild-on-error -y libxml2.spec.yaml
==> Checking for built specs at s3://olcf/ascent
==> Rebuilding libxml2@2.9.9%gcc@8.1.1~python arch=linux-rhel7-ppc64le/djuj3kq, reason: hash mismatch, remote = ivtxx4ldakqe4jxqt7cdfe7ezr5fm7d6, local = rscoqyv5fkuyzqqddmhzvkojdtledket
==> libxml2@2.9.9%gcc@8.1.1~python arch=linux-rhel7-ppc64le
    ^libiconv@1.16%gcc@8.1.1 arch=linux-rhel7-ppc64le
    ^pkgconf@1.6.3%gcc@8.1.1 arch=linux-rhel7-ppc64le
    ^xz@5.2.4%gcc@8.1.1 arch=linux-rhel7-ppc64le
    ^zlib@1.2.11%gcc@8.1.1+optimize+pic+shared arch=linux-rhel7-ppc64le
```

But `buildcache create` does not update the cache entry, giving the message `Warning: 'ParseResult' object has no attribute 's3_bucket'` shown at the bottom of this log output:
```
[lpeyrala@login1.ascent ~]$ spack -d buildcache create -afr --key prl --no-deps -d s3 -y libxml2.spec.yaml
==> [2020-01-02-11:42:34.640621] Imported buildcache from built-in commands
==> [2020-01-02-11:42:34.643427] Imported buildcache from built-in commands
==> [2020-01-02-11:42:34.644611] createtarball read spec yaml:
==> [2020-01-02-11:42:34.644644] spec:
- libxml2:
    version: 2.9.9
    arch:
      platform: linux
      platform_os: rhel7
      target: ppc64le
    compiler:
      name: gcc
      version: 8.1.1
    namespace: builtin
    parameters:
      python: false
      cflags: []
      cppflags: []
      cxxflags: []
      fflags: []
      ldflags: []
      ldlibs: []
    dependencies:
      libiconv:
        hash: mxze6sfv3o6eq4z72r674npudztt6usx
        type:
        - build
        - link
      pkgconf:
        hash: cg4qsfky7scgezvxkcdekfupatg56ues
        type:
        - build
      xz:
        hash: alc3lzqqk7zxiw7lhdgsclg5glbhxwfm
        type:
        - build
        - link
      zlib:
        hash: fmat344iyue2sfzerpscwg4lpnvg7eha
        type:
        - build
        - link
    hash: djuj3kqlwsn3t56w4dbuqcmc2zggdyyx
    build_hash: 3xjq4soyxpkw6wtep3pqvfnlf2qk2xae
- libiconv:
    version: '1.16'
    arch:
      platform: linux
      platform_os: rhel7
      target: ppc64le
    compiler:
      name: gcc
      version: 8.1.1
    namespace: builtin
    parameters:
      cflags: []
      cppflags: []
      cxxflags: []
      fflags: []
      ldflags: []
      ldlibs: []
    hash: mxze6sfv3o6eq4z72r674npudztt6usx
    build_hash: mxze6sfv3o6eq4z72r674npudztt6usx
- pkgconf:
    version: 1.6.3
    arch:
      platform: linux
      platform_os: rhel7
      target: ppc64le
    compiler:
      name: gcc
      version: 8.1.1
    namespace: builtin
    parameters:
      cflags: []
      cppflags: []
      cxxflags: []
      fflags: []
      ldflags: []
      ldlibs: []
    hash: cg4qsfky7scgezvxkcdekfupatg56ues
    build_hash: cg4qsfky7scgezvxkcdekfupatg56ues
- xz:
    version: 5.2.4
    arch:
      platform: linux
      platform_os: rhel7
      target: ppc64le
    compiler:
      name: gcc
      version: 8.1.1
    namespace: builtin
    parameters:
      cflags: []
      cppflags: []
      cxxflags: []
      fflags: []
      ldflags: []
      ldlibs: []
    hash: alc3lzqqk7zxiw7lhdgsclg5glbhxwfm
    build_hash: alc3lzqqk7zxiw7lhdgsclg5glbhxwfm
- zlib:
    version: 1.2.11
    arch:
      platform: linux
      platform_os: rhel7
      target: ppc64le
    compiler:
      name: gcc
      version: 8.1.1
    namespace: builtin
    parameters:
      optimize: true
      pic: true
      shared: true
      cflags: []
      cppflags: []
      cxxflags: []
      fflags: []
      ldflags: []
      ldlibs: []
    hash: fmat344iyue2sfzerpscwg4lpnvg7eha
    build_hash: fmat344iyue2sfzerpscwg4lpnvg7eha
==> [2020-01-02-11:42:34.666718] Reading config file /ccsopen/home/lpeyrala/.spack/linux/mirrors.yaml
==> [2020-01-02-11:42:34.668600] find_matching_specs: about to parse specs for {'/djuj3kqlwsn3t56w4dbuqcmc2zggdyyx'}
==> [2020-01-02-11:42:34.668799] Reading config file /autofs/nccsopen-svm1_home/lpeyrala/spack/etc/spack/defaults/config.yaml
==> [2020-01-02-11:42:34.686282] DATABASE LOCK TIMEOUT: 120s
==> [2020-01-02-11:42:34.686320] PACKAGE LOCK TIMEOUT: No timeout
==> [2020-01-02-11:42:34.686557] READ LOCK: /autofs/nccsopen-svm1_home/lpeyrala/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
==> [2020-01-02-11:42:34.689721] READ LOCK: /autofs/nccsopen-svm1_home/lpeyrala/spack/opt/spack/.spack-db/lock[0:0] [Acquired]
==> [2020-01-02-11:42:34.695831] READ LOCK: /autofs/nccsopen-svm1_home/lpeyrala/spack/opt/spack/.spack-db/lock[0:0] [Released]
==> [2020-01-02-11:42:34.696722] READ LOCK: /autofs/nccsopen-svm1_home/lpeyrala/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
==> [2020-01-02-11:42:34.699118] READ LOCK: /autofs/nccsopen-svm1_home/lpeyrala/spack/opt/spack/.spack-db/lock[0:0] [Acquired]
==> [2020-01-02-11:42:34.704870] READ LOCK: /autofs/nccsopen-svm1_home/lpeyrala/spack/opt/spack/.spack-db/lock[0:0] [Released]
==> [2020-01-02-11:42:34.705769] Found at least one matching spec
==> [2020-01-02-11:42:34.706267] examining match libxml2@2.9.9%gcc@8.1.1~python arch=linux-rhel7-ppc64le
==> [2020-01-02-11:42:34.706346] Reading config file /autofs/nccsopen-svm1_home/lpeyrala/spack/etc/spack/defaults/repos.yaml
==> [2020-01-02-11:42:34.773120] adding matching spec libxml2@2.9.9%gcc@8.1.1~python arch=linux-rhel7-ppc64le
==> [2020-01-02-11:42:34.773208] writing tarballs to s3://olcf/ascent/build_cache
==> [2020-01-02-11:42:35.624513] Warning: 'ParseResult' object has no attribute 's3_bucket'
==> [2020-01-02-11:42:34.644080] createtarball, reading spec from libxml2.spec.yaml
==> [2020-01-02-11:42:34.773490] creating binary cache file for package libxml2@2.9.9%gcc@8.1.1~python arch=linux-rhel7-ppc64le
```

The `ParseResult` error traces to function `remove_url` defined in `spack/lib/spack/spack/util/web.py:266` . The function `remove_url` is called on lines 327 and 344 of `binary_distribution.py` . It looks to me like the purpose here is to clean up existing, identically named binaries in the build cache, before saving the new items. If I comment out the two code blocks which attempt to preliminarily delete the existing buildcache items, and just allow the overwrite to occur, the buildcache is updated just fine, with the caveat that no attempt is made to catch NoOverwriteException. Although that works, not sure this is the best way to go about it.

**Comment by gartung** [comment]
@opadron @scottwittenburg I am not familiar with the S3 mirror part of the buildcache command. Can you comment?

**Comment by opadron** [comment]
This looks like an oversight from the latest changes with the S3 support/URL handling.

Does this change fix the issue?

`spack/lib/spack/spack/util/web.py:266`
```diff
- s3.delete_object(Bucket=url.s3_bucket, Key=url.path)
+ s3.delete_object(Bucket=url.netloc, Key=url.path)
```

**Comment by eugeneswalker** [comment]
@opadron Yes! Like a charm! Thank you Omar! Do you think we can get this merged in ASAP, perhaps for 0.14.0?

**Comment by adamjstewart** [comment]
Closed by #14412
